internal/pkg/vm/internal/vmcontext/actor_state_handle_test.go

```diff
@@ -11,120 +13,241 @@ import (
+		defer ts.cleanup()
 
-		ts.h.Take(&out)
-		ts.h.Release(&out)
-	})
+		var out testActorStateHandleState
+		ts.h.Readonly(&out)
+		ts.h.Readonly(&out)
 
-	t.Run("abort on releasing new state", func(t *testing.T) {
-		defer mustPanic(t)
+		assert.Equal(t, out, ts.initialstate)
+	})
 
+	t.Run("transaction", func(t *testing.T) {
 		ts := setup()
+		defer ts.cleanup()
 
-		var out string
-		ts.h.Take(&out)
-		ts.h.Release("some new state")
+		var out testActorStateHandleState
+		expected := "new state"
+
+		ts.h.Transaction(&out, func() (interface{}, error) {
+			// check state is not what we are going to use
+			assert.NotEqual(t, out.FieldA, expected)
+			out.FieldA = expected
+
+			return nil, nil
+		})
+		// check that it changed
+		assert.Equal(t, out.FieldA, expected)
+
+		var out2 testActorStateHandleState
+		ts.h.Readonly(&out2)
+		// really check with a new object
+		assert.Equal(t, out2.FieldA, expected)
 	})
 
-	t.Run("release and take", func(t *testing.T) {
+	t.Run("transaction but no mutation", func(t *testing.T) {
 		ts := setup()
+		defer ts.cleanup()
+
+		var out testActorStateHandleState
+
+		// should work, mutating is not compulsory
+		ts.h.Transaction(&out, func() (interface{}, error) {
+			return nil, nil
+		})
 
-		var out string
-		ts.h.Take(&out)
-		ts.h.Release(out)
-		ts.h.Take(&out)
 		assert.Equal(t, out, ts.initialstate)
 	})
 
-	t.Run("abort when calling release without taking", func(t *testing.T) {
-		defer mustPanic(t)
+	t.Run("transaction returning error", func(t *testing.T) {
+		ts := setup()
+		defer ts.cleanup()
+
+		var out testActorStateHandleState
 
+		_, err := ts.h.Transaction(&out, func() (interface{}, error) {
+			out.FieldA = "changed!"
+			return nil, testError
+		})
+		assert.Error(t, err)
+		// check that it did NOT change
+		assert.Equal(t, out, ts.initialstate)
+
+		var out2 testActorStateHandleState
+		ts.h.Readonly(&out2)
+		// really check with a new object
+		assert.Equal(t, out2, ts.initialstate)
+	})
+
+	t.Run("transaction returning value", func(t *testing.T) {
 		ts := setup()
+		defer ts.cleanup()
+
+		var out testActorStateHandleState
+
+		v, _ := ts.h.Transaction(&out, func() (interface{}, error) {
+			return out.FieldA, nil
+		})
 
-		ts.h.Release("fake")
+		assert.Equal(t, v, ts.initialstate.FieldA)
 	})
 
-	t.Run("update release", func(t *testing.T) {
-		var out string
+	t.Run("mutated after the transaction", func(t *testing.T) {
+		defer mustPanic(t)
 
 		ts := setup()
+		defer ts.cleanup()
 
-		ts.h.Take(&out)
-		// check state is not what we are going to use
-		expected := "new state"
-		assert.NotEqual(t, out, expected)
-		// update the state
-		ts.h.UpdateRelease(expected)
-		// take it out and check that we have what we put in
-		ts.h.Take(&out)
-		assert.Equal(t, out, expected)
+		var out testActorStateHandleState
+
+		ts.h.Transaction(&out, func() (interface{}, error) {
+			out.FieldA = "changed!"
+			return nil, nil
+		})
+
+		out.FieldA = "changed again!"
 	})
+}
```

Comment:
Add a test for sequential transactions, and for readonly followed by transaction.